sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/BatchSpannerRead.java

```diff
@@ -193,15 +193,16 @@ public void processElement(ProcessContext c) throws Exception {
     }
 
     private ServiceCallMetric createServiceCallMetric(
-        String projectId, String databaseId, String tableId) {
+        @Nullable String projectId, String databaseId, String tableId) {
       HashMap<String, String> baseLabels = new HashMap<>();
       baseLabels.put(MonitoringInfoConstants.Labels.PTRANSFORM, "");
       baseLabels.put(MonitoringInfoConstants.Labels.SERVICE, "Spanner");
       baseLabels.put(MonitoringInfoConstants.Labels.METHOD, "Read");
       baseLabels.put(
           MonitoringInfoConstants.Labels.RESOURCE,
           GcpResourceIdentifiers.spannerTable(projectId, databaseId, tableId));
-      baseLabels.put(MonitoringInfoConstants.Labels.SPANNER_PROJECT_ID, projectId);
+      baseLabels.put(
+          MonitoringInfoConstants.Labels.SPANNER_PROJECT_ID, (projectId == null ? "" : projectId));
```

Comment:
Is there a reason we aren't using the same project id that SpannerOptions relies on  [ServiceOptions#getDefaultProjectId](https://googleapis.dev/java/google-cloud-clients/0.97.0-alpha/com/google/cloud/ServiceOptions.Builder.html#setProjectId-java.lang.String-) if the value provider hasn't set it?

We want to make sure we keep the existing behavior of getting the default project id within the execution environment and not when the pipeline is being constructed (unless there is a strong reason for the swap).

Reply:
Agreed - I was going for the quick fix :) 

Looking further, I am also not sure why the ServiceCallMetric is being rebuilt for each element processed (instead of in `@Setup`) which seems wasteful  -- moving to `@Setup` method there in both files.

Reply:
The ServiceCallMetric was written to be per service call instance so moving it to `@Setup` will break status code reporting since the label representing the status code will be overwritten by subsequent calls. I agree that the implementation choice of combining instantiation and reporting of the status code was a poor choice but we should go back to the original implementation and address that perf issue later.

Reply:
Reverted - storing the projectID as an `@Setup` initialized member instead.